src/private_broadcast.cpp

```diff
@@ -0,0 +1,140 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://opensource.org/license/mit/.
+
+#include <private_broadcast.h>
+#include <util/check.h>
+
+/// If a transaction is not received back from the network for this duration
+/// after it is broadcast, then we consider it stale / for rebroadcasting.
+static constexpr auto STALE_DURATION{1min};
+
+bool PrivateBroadcast::Add(const CTransactionRef& tx) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    const Txid& txid = tx->GetHash();
+    LOCK(m_mutex);
+    auto [pos, inserted] = m_by_txid.emplace(txid, TxWithPriority{.tx = tx, .priority = Priority{}});
+    if (inserted) {
+        m_by_priority.emplace(Priority{}, txid);
+    }
+    return inserted;
+}
+
+std::optional<size_t> PrivateBroadcast::Remove(const CTransactionRef& tx) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    auto iters = Find(tx->GetHash());
+    if (iters && iters->by_txid->second.tx->GetWitnessHash() == tx->GetWitnessHash()) {
+        m_by_txid.erase(iters->by_txid);
+        const auto handle{m_by_priority.extract(iters->by_priority)};
+        return handle.key().num_broadcasted;
+    }
+    return std::nullopt;
+}
+
+std::optional<CTransactionRef> PrivateBroadcast::GetTxForBroadcast() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    if (!m_by_priority.empty()) {
+        const Txid& txid = m_by_priority.begin()->second;
+        auto it = m_by_txid.find(txid);
+        if (Assume(it != m_by_txid.end())) {
+            return it->second.tx;
+        }
+        m_by_priority.erase(m_by_priority.begin());
+    }
+    return std::nullopt;
+}
+
+void PrivateBroadcast::PushedToNode(const NodeId& nodeid, const Txid& txid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    m_by_nodeid.emplace(nodeid, txid);
+}
+
+std::optional<CTransactionRef> PrivateBroadcast::GetTxPushedToNode(const NodeId& nodeid) const
+    EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+
+    auto it_by_node = m_by_nodeid.find(nodeid);
+    if (it_by_node != m_by_nodeid.end()) {
+        const Txid& txid{it_by_node->second};
+        auto it_by_txid = m_by_txid.find(txid);
+        if (it_by_txid != m_by_txid.end()) {
+            return it_by_txid->second.tx;
+        }
+    }
+
+    return std::nullopt;
+}
+
+bool PrivateBroadcast::FinishBroadcast(const NodeId& nodeid, bool confirmed_by_node) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const auto handle{m_by_nodeid.extract(nodeid)};
+    if (!handle) {
+        return false;
+    }
+
+    const Txid& txid{handle.mapped()};
+
+    auto iters{Find(txid)};
+
+    if (!iters.has_value()) {
+        return false;
+    }
+
+    if (confirmed_by_node) {
+        // Update broadcast stats, since txid was found and its reception is confirmed by the node.
+        Priority& priority = iters->by_txid->second.priority;
+
+        ++priority.num_broadcasted;
+        priority.last_broadcasted = NodeClock::now();
```

Comment:
this non-purity (`NodeClock::now()`) can make testing and benchmarking a bit tougher - consider passing it as a param